[JSC] Optimize async/await and microtask queue part 1

Constellation opened the pull request:
#### ecb9f873dce9e00649d9f3d5bc668d5fd4e9bcb1
<pre>
[JSC] Optimize async/await and microtask queue part 1
<a href="https://bugs.webkit.org/show_bug.cgi?id=244165">https://bugs.webkit.org/show_bug.cgi?id=244165</a>

Reviewed by Ross Kirsling.

This patch improves our async await performance, including microtask queue implementation in JSC side.

1. Promise reaction now delivers |context| value additionally. It is passed as a second argument to the
fulfill / reject handlers if it exists. In async function code, we need to have generator reference in
the promise handlers, and we end up allocating a closure for them capturing |generator|. But (1) this
kind of case is common and (2) allocating closure is costly. In this patch, we additionally pass context
parameter so that handlers can get |generator| tied to this handler registration. This removes closure
allocations and improve async / await performance. Currently this context parameter is usable only when
using resolveWithoutPromise APIs. Keep in mind that we could attempt to optimize the current implementation
by making promiseOrCapability field to promiseOrCapabilityOrContext. So if users would like to use it,
not passing a promise as a context for the future refactoring, it could be broken.

2. We found that MicrotaskQueue can get *super* large. And current heap-allocated JSMicrotask and Strong&lt;&gt; implementation
is too costly for these cases. In this patch, we redesign MicrotaskQueue in JSC VM: QueuedTask holds JSValues directly,
and JSC GC scans this queue to keep them alive. We also make QueuedTask non heap-allocated structure so MicrotaskQueue
enqueue/dequeue and scanning (for GC) is super fast. We also introduce optimization to reduce scanning cost in GC by
maintaining scanning cursor so that we do not take much time in GC scanning. Currently these optimization is only applied
to JSC VM&apos;s MicrotaskQueue: WebCore has different implementation, so this is not applied to WebContent use case.
But we first would like to apply it to JavaScriptCore.framework use case, and then, we will apply this optimization / redesign
current WebCore MicrotaskQueue mechanism to be faster one later.

This offers 4% improvement in JetStream2/async-fs since it is using async / await.

* LayoutTests/inspector/canvas/recording-bitmaprenderer-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-bitmaprenderer-full-expected.txt:
* LayoutTests/inspector/canvas/recording-bitmaprenderer-memoryLimit-expected.txt:
* Source/JavaScriptCore/builtins/AsyncFromSyncIteratorPrototype.js:
(linkTimeConstant.asyncFromSyncIteratorOnRejected):
(linkTimeConstant.asyncFromSyncIteratorOnFulfilledContinue):
(linkTimeConstant.asyncFromSyncIteratorOnFulfilledDone):
(return):
(throw):
(next.try): Deleted.
(next): Deleted.
(return.try): Deleted.
(throw.try): Deleted.
* Source/JavaScriptCore/builtins/AsyncFunctionPrototype.js:
(linkTimeConstant.asyncFunctionResumeOnFulfilled):
(linkTimeConstant.asyncFunctionResumeOnRejected):
(linkTimeConstant.asyncFunctionResume):
* Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
(linkTimeConstant.asyncGeneratorYieldAwaited):
(linkTimeConstant.asyncGeneratorYieldOnRejected):
(linkTimeConstant.asyncGeneratorYield):
(linkTimeConstant.awaitValue):
(linkTimeConstant.doAsyncGeneratorBodyCallOnFulfilledNormal):
(linkTimeConstant.doAsyncGeneratorBodyCallOnFulfilledReturn):
(linkTimeConstant.doAsyncGeneratorBodyCall):
(linkTimeConstant.asyncGeneratorResumeNextOnFulfilled):
(linkTimeConstant.asyncGeneratorResumeNextOnRejected):
(linkTimeConstant.asyncGeneratorResumeNext):
(asyncGeneratorYieldAwaited): Deleted.
(onRejected): Deleted.
(onFulfilled): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/InternalPromiseConstructor.js:
(internalAll):
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.pushNewPromiseReaction):
(linkTimeConstant.triggerPromiseReactions):
(linkTimeConstant.promiseReactionJobWithoutPromise):
(linkTimeConstant.resolveWithoutPromise):
(linkTimeConstant.rejectWithoutPromise):
(linkTimeConstant.fulfillWithoutPromise):
(linkTimeConstant.resolveWithoutPromiseForAsyncAwait):
(linkTimeConstant.createResolvingFunctionsWithoutPromise):
(linkTimeConstant.promiseReactionJob):
(linkTimeConstant.promiseResolveThenableJobFast):
(linkTimeConstant.promiseResolveThenableJobWithoutPromiseFast):
(linkTimeConstant.promiseResolveThenableJobWithDerivedPromise):
(linkTimeConstant.performPromiseThen):
* Source/JavaScriptCore/builtins/PromisePrototype.js:
(then):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::generatorInternalFieldIndex):
(JSC::FunctionNode::emitBytecode):
* Source/JavaScriptCore/debugger/Debugger.cpp:
(JSC::Debugger::didQueueMicrotask):
(JSC::Debugger::willRunMicrotask):
(JSC::Debugger::didRunMicrotask):
* Source/JavaScriptCore/debugger/Debugger.h:
(JSC::Debugger::Observer::didQueueMicrotask):
(JSC::Debugger::Observer::willRunMicrotask):
(JSC::Debugger::Observer::didRunMicrotask):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::beginMarking):
(JSC::Heap::addCoreConstraints):
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::asyncCallIdentifier):
(Inspector::InspectorDebuggerAgent::didScheduleAsyncCall):
(Inspector::InspectorDebuggerAgent::didCancelAsyncCall):
(Inspector::InspectorDebuggerAgent::willDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didQueueMicrotask):
(Inspector::InspectorDebuggerAgent::willRunMicrotask):
(Inspector::InspectorDebuggerAgent::didRunMicrotask):
(Inspector::InspectorDebuggerAgent::didClearAsyncStackTraceData):
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
* Source/JavaScriptCore/runtime/JSGenerator.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::queueMicrotask):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runJSMicrotask):
(JSC::JSMicrotask::run):
* Source/JavaScriptCore/runtime/JSMicrotask.h:
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::performPromiseThen):
* Source/JavaScriptCore/runtime/Microtask.h:
(JSC::Microtask::Microtask):
(JSC::Microtask::identifier const):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::queueMicrotask):
(JSC::VM::drainMicrotasks):
(JSC::VM::beginMarking):
(JSC::VM::visitAggregateImpl):
(JSC::QueuedTask::run):
(JSC::MicrotaskQueue::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::QueuedTask::QueuedTask):
(JSC::QueuedTask::identifier const):
(JSC::MicrotaskQueue::dequeue):
(JSC::MicrotaskQueue::enqueue):
(JSC::MicrotaskQueue::isEmpty const):
(JSC::MicrotaskQueue::clear):
(JSC::MicrotaskQueue::beginMarking):
* Source/WTF/wtf/Deque.h:
(WTF::DequeIterator::operator+=):
(WTF::DequeIterator::operator+ const):
(WTF::DequeConstIterator::operator+=):
(WTF::DequeConstIterator::operator+ const):
(WTF::inlineCapacity&gt;::increment):
* Source/WebCore/bindings/js/JSDOMMicrotask.cpp:
(WebCore::JSDOMMicrotask::run):

Canonical link: <a href="https://commits.webkit.org/253651@main">https://commits.webkit.org/253651@main</a>
</pre>

<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a72a26f332dc05fa2615f2004c38d872885727b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30708 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17589 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95495 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149223 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90637 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29084 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25517 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78833 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90738 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92268 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23502 "Found 1 new test failure: imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-peerconnection.https.html") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73576 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23569 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78507 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/78857 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66575 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78587 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26868 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12709 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72224 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26787 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13724 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25791 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2593 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28465 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36584 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75007 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28409 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33003 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16591 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/builtins/PromiseOperations.js:
```diff
@@ -580,7 +587,7 @@ function performPromiseThen(promise, onFulfilled, onRejected, promiseOrCapabilit
     var state = flags & @promiseStateMask;
     if (state === @promiseStatePending)
-        @pushNewPromiseReaction(promise, reactionsOrResult, promiseOrCapability, onFulfilled, onRejected);
+        @pushNewPromiseReaction(promise, reactionsOrResult, promiseOrCapability, onFulfilled, onRejected, context);
```
❤️

dcrousso commented on Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
```diff
@@ -1296,36 +1296,36 @@ void InspectorDebuggerAgent::willEnter(JSC::CallFrame* callFrame)
-    int identifier = m_nextMicrotaskIdentifier++;
-    auto result = m_identifierForMicrotask.set(&microtask, identifier);
+    int callbackId = m_nextMicrotaskIdentifier++;
+    auto result = m_identifierForMicrotask.set(identifier, callbackId);
```
Since we're now using a guaranteed unique `MicrotaskIdentifier`, there's actually no reason to even have `m_nextMicrotaskIdentifier` or `m_identifierForMicrotask` anymore.

We should just be able to pass `identifier.toUInt64()` directly to `didScheduleAsyncCall` instead (though you might have to change the type from `int` to `uint64_t`).

The only reason we had `m_nextMicrotaskIdentifier` and `m_identifierForMicrotask` before was because ` JSC::Microtask` did not have a unique identifier, which the `InspectorDebuggerAgent` needs in order to keep track of which microtask was just run.

Constellation replied:
OK, changed.

rkirsling commented on Source/JavaScriptCore/builtins/PromiseOperations.js:
```diff
@@ -36,7 +36,8 @@ function pushNewPromiseReaction(thenable, existingReactions, promiseOrCapability
             @onRejected: onRejected,
-            // This is 3x then number of out of line reactions (promise, fulfill callback, reject callback).
+            @context: context,
+            // This is 4x then number of out of line reactions (promise, fulfill callback, reject callback).
```
nit: "4x then" -> "4 plus the" would be clearer

Constellation replied:
Changed.

Constellation replied:
Ah, it is `4x the number`.

rkirsling reviewed (approved):
Looks good from my perspective, but:
- For (1), I didn't understand the last sentence you wrote in the description:
    > So if users would like to use it, not passing a promise as a context for the future refactoring, it could be broken.
- For (2), I think somebody with more GC expertise should sign off too.

Constellation committed:
[JSC] Optimize async/await and microtask queue part 1
https://bugs.webkit.org/show_bug.cgi?id=244165

Reviewed by Ross Kirsling.

This patch improves our async await performance, including microtask queue implementation in JSC side.

1. Promise reaction now delivers |context| value additionally. It is passed as a second argument to the
fulfill / reject handlers if it exists. In async function code, we need to have generator reference in
the promise handlers, and we end up allocating a closure for them capturing |generator|. But (1) this
kind of case is common and (2) allocating closure is costly. In this patch, we additionally pass context
parameter so that handlers can get |generator| tied to this handler registration. This removes closure
allocations and improve async / await performance. Currently this context parameter is usable only when
using resolveWithoutPromise APIs. Keep in mind that we could attempt to optimize the current implementation
by making promiseOrCapability field to promiseOrCapabilityOrContext. So if users would like to use it,
not passing a promise as a context for the future refactoring, it could be broken.

2. We found that MicrotaskQueue can get *super* large. And current heap-allocated JSMicrotask and Strong<> implementation
is too costly for these cases. In this patch, we redesign MicrotaskQueue in JSC VM: QueuedTask holds JSValues directly,
and JSC GC scans this queue to keep them alive. We also make QueuedTask non heap-allocated structure so MicrotaskQueue
enqueue/dequeue and scanning (for GC) is super fast. We also introduce optimization to reduce scanning cost in GC by
maintaining scanning cursor so that we do not take much time in GC scanning. Currently these optimization is only applied
to JSC VM's MicrotaskQueue: WebCore has different implementation, so this is not applied to WebContent use case.
But we first would like to apply it to JavaScriptCore.framework use case, and then, we will apply this optimization / redesign
current WebCore MicrotaskQueue mechanism to be faster one later.

This offers 4% improvement in JetStream2/async-fs since it is using async / await.

* LayoutTests/inspector/canvas/recording-bitmaprenderer-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-bitmaprenderer-full-expected.txt:
* LayoutTests/inspector/canvas/recording-bitmaprenderer-memoryLimit-expected.txt:
* Source/JavaScriptCore/builtins/AsyncFromSyncIteratorPrototype.js:
(linkTimeConstant.asyncFromSyncIteratorOnRejected):
(linkTimeConstant.asyncFromSyncIteratorOnFulfilledContinue):
(linkTimeConstant.asyncFromSyncIteratorOnFulfilledDone):
(return):
(throw):
(next.try): Deleted.
(next): Deleted.
(return.try): Deleted.
(throw.try): Deleted.
* Source/JavaScriptCore/builtins/AsyncFunctionPrototype.js:
(linkTimeConstant.asyncFunctionResumeOnFulfilled):
(linkTimeConstant.asyncFunctionResumeOnRejected):
(linkTimeConstant.asyncFunctionResume):
* Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
(linkTimeConstant.asyncGeneratorYieldAwaited):
(linkTimeConstant.asyncGeneratorYieldOnRejected):
(linkTimeConstant.asyncGeneratorYield):
(linkTimeConstant.awaitValue):
(linkTimeConstant.doAsyncGeneratorBodyCallOnFulfilledNormal):
(linkTimeConstant.doAsyncGeneratorBodyCallOnFulfilledReturn):
(linkTimeConstant.doAsyncGeneratorBodyCall):
(linkTimeConstant.asyncGeneratorResumeNextOnFulfilled):
(linkTimeConstant.asyncGeneratorResumeNextOnRejected):
(linkTimeConstant.asyncGeneratorResumeNext):
(asyncGeneratorYieldAwaited): Deleted.
(onRejected): Deleted.
(onFulfilled): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/InternalPromiseConstructor.js:
(internalAll):
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.pushNewPromiseReaction):
(linkTimeConstant.triggerPromiseReactions):
(linkTimeConstant.promiseReactionJobWithoutPromise):
(linkTimeConstant.resolveWithoutPromise):
(linkTimeConstant.rejectWithoutPromise):
(linkTimeConstant.fulfillWithoutPromise):
(linkTimeConstant.resolveWithoutPromiseForAsyncAwait):
(linkTimeConstant.createResolvingFunctionsWithoutPromise):
(linkTimeConstant.promiseReactionJob):
(linkTimeConstant.promiseResolveThenableJobFast):
(linkTimeConstant.promiseResolveThenableJobWithoutPromiseFast):
(linkTimeConstant.promiseResolveThenableJobWithDerivedPromise):
(linkTimeConstant.performPromiseThen):
* Source/JavaScriptCore/builtins/PromisePrototype.js:
(then):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::generatorInternalFieldIndex):
(JSC::FunctionNode::emitBytecode):
* Source/JavaScriptCore/debugger/Debugger.cpp:
(JSC::Debugger::didQueueMicrotask):
(JSC::Debugger::willRunMicrotask):
(JSC::Debugger::didRunMicrotask):
* Source/JavaScriptCore/debugger/Debugger.h:
(JSC::Debugger::Observer::didQueueMicrotask):
(JSC::Debugger::Observer::willRunMicrotask):
(JSC::Debugger::Observer::didRunMicrotask):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::beginMarking):
(JSC::Heap::addCoreConstraints):
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::asyncCallIdentifier):
(Inspector::InspectorDebuggerAgent::didScheduleAsyncCall):
(Inspector::InspectorDebuggerAgent::didCancelAsyncCall):
(Inspector::InspectorDebuggerAgent::willDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didQueueMicrotask):
(Inspector::InspectorDebuggerAgent::willRunMicrotask):
(Inspector::InspectorDebuggerAgent::didRunMicrotask):
(Inspector::InspectorDebuggerAgent::didClearAsyncStackTraceData):
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
* Source/JavaScriptCore/runtime/JSGenerator.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::queueMicrotask):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runJSMicrotask):
(JSC::JSMicrotask::run):
* Source/JavaScriptCore/runtime/JSMicrotask.h:
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::performPromiseThen):
* Source/JavaScriptCore/runtime/Microtask.h:
(JSC::Microtask::Microtask):
(JSC::Microtask::identifier const):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::queueMicrotask):
(JSC::VM::drainMicrotasks):
(JSC::VM::beginMarking):
(JSC::VM::visitAggregateImpl):
(JSC::QueuedTask::run):
(JSC::MicrotaskQueue::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::QueuedTask::QueuedTask):
(JSC::QueuedTask::identifier const):
(JSC::MicrotaskQueue::dequeue):
(JSC::MicrotaskQueue::enqueue):
(JSC::MicrotaskQueue::isEmpty const):
(JSC::MicrotaskQueue::clear):
(JSC::MicrotaskQueue::beginMarking):
* Source/WTF/wtf/Deque.h:
(WTF::DequeIterator::operator+=):
(WTF::DequeIterator::operator+ const):
(WTF::DequeConstIterator::operator+=):
(WTF::DequeConstIterator::operator+ const):
(WTF::inlineCapacity>::increment):
* Source/WebCore/bindings/js/JSDOMMicrotask.cpp:
(WebCore::JSDOMMicrotask::run):

Canonical link: https://commits.webkit.org/253651@main
